File: cpp/LeetCode/No804.cpp
#include <iostream>
#include <vector>
#include <set>

using namespace std;

class Solution {
public:
    int uniqueMorseRepresentations(vector<string>& words) {
        set<string> mem;
        for(string word : words){
            mem.insert(strToMorse(word));
        }
        return mem.size();
    }
private:
    vector<string> morseTable = {".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."};
    string strToMorse(string word){
        string res = "";
        for(char c : word){
            res.append(morseTable[c-'a']);
        }
        return res;
    }
};

int main(void){
    vector<string> words = {"gin","zen","gig","msg"};
    Solution solution;
    cout << solution.uniqueMorseRepresentations(words) << endl;
    return 0;
}

File: python/LeetCode Quiz/No804.py
from typing import List

class Solution:
    _morseTable = [".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."]
    def uniqueMorseRepresentations(self, words: List[str]) -> int:
        def strToMorse(word: str) -> str:
            res = ''
            for c in word:
                res+=self._morseTable[ord(c)-ord('a')]
            return res
        mem = set()
        for word in words:
            mem.add(strToMorse(word))
        return len(mem)

def main():
    words = ["gin","zen","gig","msg"]
    print(Solution.uniqueMorseRepresentations(Solution(), words))

if __name__=='__main__':
    main()